PMM-5761 indexstats collector

percona-csalguero opened the pull request:
This branch is based on metrics backward compatibility because I needed the `walkTo` func implemented in that branch.

cfsalguero committed:
PMM-5756 Compatibility code

It is still WIP. The code to rename a metric to make it compatible is
done but the compatibility list has only one metric

cfsalguero committed:
WIP

cfsalguero committed:
WIP

cfsalguero committed:
WIP

cfsalguero committed:
Merge branch 'master' into PMM-5756-metrics_backward_compatibility

cfsalguero committed:
PMM-5756 v1 Metrics compatibility

In order to expose v1 compstible metrics, I added conversion functions.
A new metric like: mongodb_ss_asserts{assert_type=*} will be converted
to the old-compatible metric:
mongodb_mongod_asserts_total{type="regular|warning|msg|user|rollovers"}.

In this particular case, conversion would be:
  conversion {
   newName: "mongodb_ss_asserts",
   oldName: "mongodb_mongod_asserts_total",
   labels : map[string]string{ "assert_type": "type"},
  }.

Some other metric renaming are more complex.
In some cases, there is a total renaming, with new labels and the only
part we can use to identify a metric is its prefix. Example:
Metrics like:
`mongodb_ss_metrics_operation_fastmod` or
`mongodb_ss_metrics_operation_idhack` or
`mongodb_ss_metrics_operation_scanAndOrder`
should use the trimmed "prefix" mongodb_ss_metrics_operation from the
metric name, and that remaining suffix is the label value for a new label
"prefixLabel".
It means that the metric (current) mongodb_ss_metrics_operation_idhack will
become into the old equivalent one :
`mongodb_mongod_metrics_operation_total {"state": "idhack"}` as defined in
the conversion slice:
```
   {
     oldName:     "mongodb_mongod_metrics_operation_total", //{state="fastmod|idhack|scan_and_order"}
     prefix:      "mongodb_ss_metrics_operation",           // _[fastmod|idhack|scanAndOrder]
     prefixLabel: "state",
   },

suffixMapping field:
--------------------
Also, some metrics suffixes for the second renaming case need a mapping betwwen the old and new values.
For example, the metric mongodb_ss_wt_cache_bytes_currently_in_the_cache has mongodb_ss_wt_cache_bytes
as the prefix so the suffix is bytes_currently_in_the_cache should be converted to a mertic named
mongodb_mongod_wiredtiger_cache_bytes and the suffix bytes_currently_in_the_cache is being mapped to
"total".

Locks are kind of special metrics so there is a special func to add them
to the results.

```

cfsalguero committed:
Merge branch 'master' into PMM-5756-metrics_backward_compatibility

cfsalguero committed:
PMM-5756 Updated deps

cfsalguero committed:
WIP before rebase

cfsalguero committed:
PMM-5761 Implemented indexStats collector

cfsalguero committed:
PMM-5756 Fixed lock metrics error

cfsalguero committed:
Merge branch 'PMM-5756-metrics_backward_compatibility' into PMM-5761_indexstats_collector

cfsalguero committed:
PMM-5756 Updated for CR

cfsalguero committed:
Merge branch 'PMM-5756-metrics_backward_compatibility' into PMM-5761_indexstats_collector

cfsalguero committed:
PMM-5761 Updated deps

askomorokhov commented on exporter/indexstats_collector.go:
```diff
@@ -0,0 +1,110 @@
+}
+
+func (d *indexstatsCollector) Collect(ch chan<- prometheus.Metric) {
+	ctx := context.TODO()
```
It is not clear for me why and how long we should use `TODO()` context here.
Could you please add a comment.

BupycHuk replied:
+1

askomorokhov reviewed (approved):
Looks good to me with comment. 👍

BupycHuk commented on exporter/indexstats_collector.go:
```diff
@@ -0,0 +1,110 @@
+			continue
+		}
+
+		buf, _ := json.MarshalIndent(stats, "", "  ")
```
looks different from what you did in #18

BupycHuk commented on exporter/indexstats_collector.go:
```diff
@@ -0,0 +1,110 @@
+		ops = val
+	}
+	if val, ok := walkTo(m, []string{"building"}).(float64); ok {
+		ops = val
```
```suggestion
		building = val
```

BupycHuk commented on exporter/indexstats_collector_test.go:
```diff
@@ -0,0 +1,107 @@
+	}
+
+	// The last \n at the end of this string is important
+	expected := strings.NewReader(`
```
Why all values here are zero? Can we add something to change these values?

BupycHuk replied:
If we can't let's add one more test for `sanitizeMetrics`

cfsalguero committed:
PMM-5756 Fixes for CR

cfsalguero committed:
Merge branch 'PMM-5756-metrics_backward_compatibility' into PMM-5761_indexstats_collector

cfsalguero committed:
PMM-5671 Fixes for CR

- Added a new test for sanitizeMetrics
- New asFloat64 func to avoid repeating code

BupycHuk commented on exporter/indexstats_collector_test.go:
```diff
@@ -0,0 +1,152 @@
+		}
+		want := primitive.M{
+			"accesses": primitive.M{
+				"ops": float64(0),
```
why values are 0 if it's 3 in input?

percona-csalguero replied:
Fixed

BupycHuk reviewed (changes requested):
Looks like something wrong in `TestSanitize`

cfsalguero committed:
PMM-5761 Fixed sanitize metrics

BupycHuk reviewed (changes requested):
Please fix merge conflicts

cfsalguero committed:
PMM-5761 Fixed merge conflicts
